### PR TITLE
WAF: Introduce logging when a request is blocked

### PR DESCRIPTION
**Pull request by miguelxpn**
#### Changes proposed in this Pull Request:

* Logs to file and to database when a request is blocked.

#### Jetpack product discussion

1201069996155224-as-1201965730420999

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Checkout this branch
* Build the plugin with it's dependencies
* Activate it on your test environment
* Make sure that the WAF log directory was created
* Test a blocked request and verify that it was logged in the file and in the database.

### CODE REVIEW

File: projects/packages/waf/src/class-waf-runtime.php
Comment (singerb) [inline projects/packages/waf/src/class-waf-runtime.php:266]: The Scan side will delete this file once ingested, right? I still wonder if we should add a filesize check here just so we don't kill their FS space if the remote scan isn't working.

**Review by singerb — COMMENTED** [review]
Is the handling on the remote scan side to ingest these still in place? In both the old and new architectures?

File: projects/packages/waf/src/class-waf-runtime.php
Comment (miguelxpn) [reply projects/packages/waf/src/class-waf-runtime.php:266]: Yep, but a check seems like a reasonable safety step. I'll add it. Any suggestions on a max size?

Comment (singerb) [reply projects/packages/waf/src/class-waf-runtime.php:266]: Not sure; I'd see if you get a sense of what the average per message is over, say 100 block log lines, and then set it at a level where we hopefully won't lose log data daily on busy sites, but won't use up all the space on a shared host site. 1GB? 500MB?

Comment (miguelxpn) [reply projects/packages/waf/src/class-waf-runtime.php]: I added a check. 100 mb to start with, we can always adjust later if we think it's not enough.

File: projects/packages/waf/src/class-waf-runtime.php
Comment (singerb) [inline projects/packages/waf/src/class-waf-runtime.php:266]: Missing a check here.

**Review by singerb — CHANGES_REQUESTED** [review]
This fataled for me when blocking a request:

```
[Thu Apr 14 17:41:15.143841 2022] [php:notice] [pid 397] [client 172.19.0.1:41656] /var/www/html/wp-content/jetpack-waf/waf-blocklog
[Thu Apr 14 17:41:15.144059 2022] [php:error] [pid 397] [client 172.19.0.1:41656] PHP Fatal error:  Uncaught TypeError: fwrite(): Argument #1 ($stream) must be of type resource, bool given in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php:269\nStack trace:\n#0 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php(269): fwrite()\n#1 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php(241): Automattic\\Jetpack\\Waf\\Waf_Runtime->write_blocklog()\n#2 /usr/local/src/jetpack-monorepo/projects/packages/waf/rules/rules.php(13): Automattic\\Jetpack\\Waf\\Waf_Runtime->block()\n#3 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runner.php(94): include('...')\n#4 /usr/local/src/jetpack-monorepo/projects/packages/waf/run.php(18): Automattic\\Jetpack\\Waf\\Waf_Runner::run()\n#5 /usr/local/src/jetpack-monorepo/projects/packages/waf/actions.php(41): require_once('...')\n#6 /var/www/html/wp-includes/class-wp-hook.php(307): Automattic\\Jetpack\\Autoloader\\jpf11009ded9fc4592b6a05b61ce272b3c_jetpack\xe2\x93\xa510_9_a_2\\Version_Loader->Automattic\\Jetpack\\Waf\\{closure}()\n#7 /var/www/html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()\n#8 /var/www/html/wp-includes/plugin.php(474): WP_Hook->do_action()\n#9 /var/www/html/wp-settings.php(427): do_action()\n#10 /var/www/html/wp-config.php(109): require_once('...')\n#11 /var/www/html/wp-load.php(50): require_once('...')\n#12 /var/www/html/wp-blog-header.php(13): require_once('...')\n#13 /var/www/html/index.php(17): require('...')\n#14 {main}\n\nNext TypeError: fclose(): Argument #1 ($stream) must be of type resource, bool given in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php:271\nStack trace:\n#0 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php(271): fclose()\n#1 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php(241): Automattic\\Jetpack\\Waf\\Waf_Runtime->write_blocklog()\n#2 /usr/local/src/jetpack-monorepo/projects/packages/waf/rules/rules.php(13): Automattic\\Jetpack\\Waf\\Waf_Runtime->block()\n#3 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runner.php(94): include('...')\n#4 /usr/local/src/jetpack-monorepo/projects/packages/waf/run.php(18): Automattic\\Jetpack\\Waf\\Waf_Runner::run()\n#5 /usr/local/src/jetpack-monorepo/projects/packages/waf/actions.php(41): require_once('...')\n#6 /var/www/html/wp-includes/class-wp-hook.php(307): Automattic\\Jetpack\\Autoloader\\jpf11009ded9fc4592b6a05b61ce272b3c_jetpack\xe2\x93\xa510_9_a_2\\Version_Loader->Automattic\\Jetpack\\Waf\\{closure}()\n#7 /var/www/html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()\n#8 /var/www/html/wp-includes/plugin.php(474): WP_Hook->do_action()\n#9 /var/www/html/wp-settings.php(427): do_action()\n#10 /var/www/html/wp-config.php(109): require_once('...')\n#11 /var/www/html/wp-load.php(50): require_once('...')\n#12 /var/www/html/wp-blog-header.php(13): require_once('...')\n#13 /var/www/html/index.php(17): require('...')\n#14 {main}\n  thrown in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runtime.php on line 271
```

We should check the return from `fopen`, but also figure out why it was failing to begin with. This was in the docker env; I added the first error log to show what file it was trying to open: `/var/www/html/wp-content/jetpack-waf/waf-blocklog`

File: projects/packages/waf/src/class-waf-runner.php
Comment (singerb) [inline projects/packages/waf/src/class-waf-runner.php:205]: Some merge issues here?

**Review by singerb — APPROVED** [review]
Doesn't fatal for me any more, although it still doesn't write the log file. Conditional approval once the merge issues are fixed.

File: projects/packages/waf/src/class-waf-runner.php
Comment (miguelxpn) [reply projects/packages/waf/src/class-waf-runner.php]: Yep, tried to resolve them using the web interface and probably messed it up somehow. Fixed it.

**Review by singerb — APPROVED** [review]
Approving now those are fixed.

**Pull request merged**
